pkg/endpointslice/map.go

```diff
@@ -0,0 +1,75 @@
+package endpointslice
+
+import (
+	"sync"
+
+	"github.com/submariner-io/lighthouse/pkg/constants"
+	discovery "k8s.io/api/discovery/v1beta1"
+)
+
+type endpointInfo struct {
+	key     string
+	hostIPs map[string][]string
+}
+
+type Map struct {
+	epMap map[string]*endpointInfo
+	sync.RWMutex
+}
+
+func (m *Map) GetIPs(hostname, cluster, namespace, name string) ([]string, bool) {
+	key := keyFunc(namespace, name, cluster)
+
+	m.RLock()
+	defer m.RUnlock()
+
+	result, ok := m.epMap[key]
+	if !ok {
+		return nil, false
+	}
+
+	ips, ok := result.hostIPs[hostname]
+
+	return ips, ok
+}
+
+func NewMap() *Map {
+	return &Map{
+		epMap: make(map[string]*endpointInfo),
+	}
+}
+
+func (m *Map) Put(es *discovery.EndpointSlice) {
+	key, ok := es.Labels[constants.LabelServiceImportName]
+	if !ok {
+		return
+	}
+
+	epInfo := &endpointInfo{
+		key:     key,
+		hostIPs: make(map[string][]string),
+	}
+
+	m.Lock()
+	defer m.Unlock()
```

Comment:
This could be moved even further to right before L62. We don't need to hold the lock which building the `epInfo` struct.